Geometry 1D modifications

tbody-cfs opened the pull request:
*n.b. feel free to edit/comment if there are errors in the logic presented here*

# MK edit

This PR aims to significantly improve how we define the geometry of our 1D simulations.
It adds the following functionality:

- Ability to define input file parameters in terms of Lpar, including flux expansion
- Ability to define a profile of magnetic pitch
- Ability to define a profile of SOL broadening
- Calculation of flux tube radial cross-sectional area for communication with reservoirs
- Passing all of the above into the state for use in other components

# Accounting for cross-field transport and flux-expansion in 1D simulations

Simple models such as the extended Lengyel and DLS models show the importance of capturing cross-field transport (i.e. $\lambda_{INT}=\lambda_q + 1.64S$) and flux expansion (i.e. $B$). We want to be able to simply vary these parameters to study their effect and compare Hermes-3 to other models.

One implementation we might want to use as a reference is [Derks et al., 2024](https://doi.org/10.1088/1361-6587/ad2e37). In this paper, conservation of heat flux is used to give
$$\lambda(x) \cdot \sin(\theta)(x) \cdot B_{trans} = const.$$
where $\lambda$ is the poloidal width of the heat flux channel and $\sin(\theta)=B_{pol}/B_{tor}$. Let's introduce the shorthand $\sin(\theta)=\Theta=B_{pol}/B_{tor}$.

$B_{trans}$ is a modification to the actual magnetic field $B_{field}$ intended to mock up cross-field transport, such that the magnetic field in the model is set as $B_{model}=B_{field} \cdot B_{trans}$.

## Discussion with @bendudson 

In 1D, $J \cdot dx \cdot dy \cdot dz = V_{cell}$

where $dx = 1$ and $dz = 2 \pi$ by default.

These coordinates are set in https://github.com/boutproject/hermes-3/blob/master/hermes-3.cxx#L247.

Looking at this, we essentially can vary $J$ to set $A_\perp$.

$$J = 1/B_{model} = b / B_{actual}$$
would do what we're trying to achieve

## Final implementation

**Functionality of the `fieldline_geometry` Component**

The `fieldline_geometry` component in Hermes-3 is designed to handle the geometric calculations necessary for 1D simulations, particularly those aligned with magnetic field lines.

**Key Responsibilities:**

* **Parallel Length Calculation:** Computes the parallel distance (`lpar`) from an upstream reference point along the magnetic field line. This is a fundamental quantity for 1D simulations as it defines the spatial coordinate.
* **Geometric Quantities:** Calculates essential geometric quantities that vary along the field line:
    * Flux tube width.
    * Flux expansion.
    * Cell dimensions (poloidal length, side area, and volume).
* **Magnetic Field Handling:**
    * Accepts user-defined expressions for the poloidal (`Bpol`) and toroidal (`Btor`) components of the magnetic field.
    * Provides an option to compute the toroidal field component from the major radius (`R`), which is a common simplification in some contexts.
    * Calculates the total magnetic field strength (`B`) and the pitch angle (`\\(\\theta\\)`).
* **Jacobian Calculation:** Computes the Jacobian (`J`), which is crucial for coordinate transformations and volume integrals in the simulation.
* **Output:** Adds the calculated geometric quantities to the simulation state for use by other components and for diagnostic output.

**Equations:**

* Pitch angle:  $\sin(\theta) = B_{pol} / B$
* Toroidal magnetic field (if computed): $B_{tor} = B_{tor,up} R_{up} / R$
* Flux expansion: $f_{exp} = (B_{pol,up} / B_{up}) / (B_{pol} / B)$
* Flux tube width: $\lambda_{int} f_{exp}$
* Cell poloidal length: $dl_{pol} = dl_\parallel B_{pol} / B$
* Cell side area: $2 \pi R \cdot dl_{pol}$
* Cell volume: $2 \pi R \cdot dl_{pol} \lambda_{int} f_{exp}$
* Jacobian: $J = (\lambda_{int} / \lambda_{q,up}) / B$

**In essence:**

The `fieldline_geometry` component provides the geometric foundation upon which 1D field-aligned simulations in Hermes-3 are built. It handles the complexities of the magnetic geometry, allowing other components to focus on the physics of the plasma.

tbody-cfs committed:
Initial work on 1D geometry

tbody-cfs committed:
Reformulate cross-field transport

tbody-cfs committed:
Tidy up

tbody-cfs committed:
Set default to no broadening

tbody-cfs commented:
# Test
![broadening](https://github.com/user-attachments/assets/10703713-3dd1-4223-b31d-23c549900d9a)

[evidence.zip](https://github.com/user-attachments/files/19980119/evidence.zip)

tbody-cfs committed:
Start documentation for geometry 1D

tbody-cfs committed:
Calculate geometry terms needed for other functions

tbody-cfs committed:
Fixes for fieldline_geometry

tbody-cfs commented:
One thing I've found — if the magnetic field doesn't equal the jacobian, we end up with a non-1 Jacobian. This has a small but noticeable impact on the simulation. Should the Jacobian be normalized such that it is 1 if the magnetic field is constant?
![image](https://github.com/user-attachments/assets/41d174ae-af6f-41de-a550-24dd2bfb1006)

tbody-cfs commented:
Full round of tests [test_fieldline_geometry.pdf](https://github.com/user-attachments/files/20015505/test_fieldline_geometry.pdf)

tbody-cfs committed:
More intuitive input variables

tbody-cfs committed:
Update documentation

mikekryjak reviewed (changes requested):
Hi @**Tom Body**, really great work on this - very nice clear code with lots of comments, and I really appreciate the docs as well as the test pdf you sent. Once this is merged, I think we should calculate some similar variables for 2D as well, that's been on my list for ages.

My comments are mostly about diagnostic naming and documentation location. 

I think it would also be really good if we had a proper test as we discussed at the dev meeting last week. Do you think this is something you have time to do? If not, then could you make a test by slapping some asserts in a Python script which will read the non-trivial formulation case?

tbody-cfs committed:
Address review comments: rename diagnostic output variables

- pitch_angle -> magnetic_pitch (sin(θ) is not an angle)
- fieldline_radius -> Rxy (BOUT++ nomenclature)
- poloidal/toroidal/total_magnetic_field -> Bpxy/Btxy/Bxy (BOUT++ nomenclature)
- flux_expansion -> f_R (disambiguation from poloidal flux expansion f_x)
- cell_poloidal_length -> dlpol (consistent with dlpar naming)
- Move fieldline_geometry docs to after domain_grid in toctree

tbody-cfs committed:
Fix build

tbody-cfs committed:
Add unit tests for fieldline_geometry

tbody-cfs committed:
Run prek tool

mikekryjak commented on src/fieldline_geometry.cxx:
```diff
@@ -0,0 +1,319 @@
+      if (j > 0) {
+        // Calculate lpar at the current cell center by adding the average of the current
+        // and previous cell widths to the previous lpar value. This is a simple trapezoidal rule.
+        lpar(0, j, 0) = lpar(0, j - 1, 0) + 0.5 * dy(0, j - 1, 0) + 0.5 * dy(0, j, 0);
```
Have you tested this when running in parallel? This looks reasonable on a single rank, but presumably each rank after the first one would need to accumulate on the connection length from the previous ranks. But I don't see you using `id` for anything in this loop.

tbody-cfs replied:
Good catch — I tested this in the Docker image, running a non-trivial 1D case at 1, 4 and 8 ranks and comparing both the geometry outputs and the evolved solution cell-by-cell against the single-rank run. It surfaced two genuine parallel bugs.

On the loop itself / id: the accumulation is correct across ranks, but you're right that id wasn't doing anything. The mechanism is that each mesh->communicate(lpar) carries the running integral one rank downstream, so repeating the integrate-then-communicate step NYPE times propagates the value to every rank. id was just the repeat counter — I've renamed it sweep and documented that. I also fixed an off-by-one out-of-bounds write in the inner loop (j <= mesh->LocalNy → j < mesh->LocalNy). With this, lpar matches the single-rank result to machine precision.

Bug 1 — upstream reference used each rank's local first cell. Several quantities referenced field(0, mesh->ystart, 0), which is the local first cell, not the global upstream. So f_R, transport_broadening and Btor (with compute_Btor_from_R) — and hence the Jacobian — were only right on rank 0; every other rank restarted f_R at 1.0 at its local boundary:

f_R, np=4, first cells of each rank vs single-rank reference:
rank 0: [1.000 1.000 1.001 1.003]  ref [1.000 1.000 1.001 1.003]  ✓
rank 1: [1.000 0.997 0.994 0.991]  ref [1.849 1.843 1.837 1.830]  ✗
rank 2: [1.000 0.997 0.994 0.991]  ref [1.200 1.197 1.193 1.189]  ✗
rank 3: [1.000 1.000 0.999 0.998]  ref [1.016 1.016 1.015 1.015]  ✗

Fixed with an upstream_value() helper that reads the reference on rank 0 and MPI_Bcasts it. After this, all geometry outputs (lpar, Rxy, lambda_int, f_R, dlpol, Bxy) match the serial run to machine precision (max rel diff 0.0) at np = 1/4/8.

Bug 2 — coord->J was set only on interior cells. The geometry outputs matching wasn't sufficient: comparing the evolved solution showed a larger np1-vs-np4 spread for the default 1d-threshold case. The Jacobian was written only for ystart..yend, so the guard cells kept the default mesh J, discontinuous with the fieldline J. I now set J over the whole local field (guard cells included) and mesh->communicate(coord->J), mirroring recalculate_metric. Max relative np1-vs-np4 difference at a fixed output step:

field   before J-fix   after J-fix   1d-threshold
Ne         1.30%          0.57%          0.57%
Td         2.49%          1.49%          1.47%
Nd         1.60%          0.80%          0.80%

i.e. the fix brings the parallel spread down to the baseline the 1d-threshold already finds. Added a unit-test assertion that J is set in the guard cells (it fails against the old interior-only code).

On the residual difference: the ~0.5–1% that remains is not geometry-specific — the 1D-threshold shows the same, because this is the beuler steady-state solver at a fixed output step (snes_local_residual ~ 1e-3); the two decompositions take slightly different internal steps and land at slightly different points on a still-evolving transient. Driven to a tighter residual they converge to the same steady state.

mikekryjak commented on src/fieldline_geometry.cxx:
```diff
@@ -0,0 +1,319 @@
+  // Poloidal area of the cell (poloidal length times circumference)
+  cell_side_area = cell_poloidal_length * 2.0 * PI * fieldline_radius;
+  // Volume of the cell (poloidal area times flux tube width)
+  cell_volume = cell_side_area * flux_tube_width;
```
I see that none of the new geometry variables are put into the state for other components to read. Is this intentional? An easy way to do this would be to add code in `transform_impl` to put them into the state at every RHS. A better way would be to do it in the `coordinates` object, but that would be a BOUT++ change and it would be more complicated. I personally think the former is sufficient.

I know this is already being called by the new reservoirs - how does that work? @PoloidalLloyd

tbody-cfs replied:
Not intentional — good suggestion, and I've taken the transform_impl approach you describe (agree the coordinates-object route is more invasive than it's worth here). transform_impl now publishes the geometry quantities into the shared state on every RHS, with matching readWrite permissions declared in the constructor, so other components can get<Field3D>() them. The keys mirror the diagnostic names — fieldline_geometry_Rxy, _Bxy, _lpar, _dlpol, _cell_volume, etc. Added a unit test (TransformPublishesGeometryToState) that checks all of them are published.

On the reservoirs — I don't have that component on this branch, so nothing consumes these yet here; this just makes them available. @PoloidalLloyd, could you confirm which keys the reservoir reads so I can make sure the names line up? Happy to rename to match whatever it expects.

mikekryjak replied:
I talked to Lloyd and he will raise a separate PR to connect the reservoirs up.

mikekryjak commented on src/fieldline_geometry.cxx:
```diff
@@ -0,0 +1,319 @@
+  // Bxy is no longer consistent with the Jacobian.
+  // Set equal to NaN, to prevent anyone from using it.
+  BOUT_FOR(i, coord->Bxy.getRegion("RGN_ALL")) {
+    coord->Bxy[i] = std::numeric_limits<BoutReal>::quiet_NaN();
```
Hmm... I think `Bxy` almost certainly isn't used for any physics relevant to 1D simulations, but not 100% sure. @bendudson what do you think?

tbody-cfs replied:
Looks like Bxy consumer is used in perpendicular-dynamics / turbulence / EM components (vorticity, the perpendicular operators in div_ops, diamagnetic & polarisation drifts, classical + ion/electron viscosity, neutral_full_velocity, electromagnetic, and the Alfvén-speed term in sound_speed)

mikekryjak reviewed (changes requested):
This is great work @tbody-cfs, especially the tests! I have a few questions, see below.

tbody-cfs committed:
Respond to review

tbody-cfs committed:
Fixes due to updated BOUT version

tbody-cfs committed:
Fix "Option fieldline_geometry:lambda_int has no value" thrown in the test body."

mikekryjak reviewed (approved):
Good to go - thanks @tbody-cfs!
